src/contexts/authContext.js

```diff
@@ -0,0 +1,113 @@
+import { createContext, useContext, useEffect, useState } from 'react';
+import { Auth, Hub } from 'aws-amplify';
+import { CognitoHostedUIIdentityProvider } from '@aws-amplify/auth';
+import { useHistory } from 'react-router';
+
+const AuthContext = createContext();
+
+const AuthProvider = ({ children }) => {
+  const [user, setUser] = useState({ data: null, loading: true });
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState(null);
+  const history = useHistory();
+
+  const asyncHandler =
+    (fn) =>
+    (...args) => {
+      setLoading(true);
+      setError(null);
+      fn(...args)
+        .catch(setError)
+        .finally(() => {
+          setLoading(false);
+        });
+    };
+
+  const signup = asyncHandler(async (name, email, password) => {
+    await Auth.signUp({
+      username: email,
+      password,
+      attributes: {
+        name,
+        email,
+      },
+    });
+    history.push('/verify-account', {
+      email,
+      password,
+    });
+  });
+
+  const login = asyncHandler(async (email, password) => {
+    await Auth.signIn(email, password);
+  });
+
+  const confirmSignup = asyncHandler(async (email, password, code) => {
+    await Auth.confirmSignUp(email, code);
+    await login(email, password);
+  });
+
+  const logout = asyncHandler(async () => {
+    await Auth.signOut();
+  });
+
+  const loginProvider = asyncHandler(async (provider) => {
+    await Auth.federatedSignIn({
+      provider: CognitoHostedUIIdentityProvider[provider],
+    });
+  });
+
+  const getUser = () => {
+    Auth.currentAuthenticatedUser()
+      .then((currentUser) => {
+        setUser({
+          data: currentUser.attributes,
+          loading: false,
+        });
+      })
+      .catch(() => {
+        setUser({ data: null, loading: false });
+      });
+  };
+
+  useEffect(() => {
+    const unsubscribe = Hub.listen('auth', ({ payload: { event, data } }) => {
+      switch (event) {
+        case 'signIn':
+          getUser();
+          break;
+        case 'signOut':
+          setUser({ data: null, loading: false });
+          break;
+        default:
+      }
+    });
+
+    getUser();
+    return unsubscribe;
+  }, []);
+
+  const value = {
+    user,
+    loading,
+    error,
+    signup,
+    confirmSignup,
+    login,
+    logout,
+    loginProvider,
+  };
```

Comment:
I recommend to use `useMemo` to wrap the context value hook in order to avoid unnecessary re-renders on any child components. This is the explanation:

_Everytime the `AuthProvider` component re-renders a new reference to **object** or **array** is being passed as `value` to `AuthContext.Provider` and hence even if the actual value may not have changed, the Context consumers are re-rendered since the reference check fails for the value._

_Now you may or may not need a `useMemo` depending on what logic you have in your **ContextProvider**. For instance, in your case the `AuthContext` is just using one state and passes it on to the child and if the `AuthContext` is one of the top level element which is not re-rendered by anything other than the value change then in that case whether you use `useMemo` or not makes no difference since the reference of the returned value from `useMemo` is also updated on value change._

_However if you have certain parents to `AuthProvider` which can cause `AuthProvider` to re-render, `useMemo` to memoize context value comes in handy to avoid re-rendering of all the context consumers._